File: 프로그래머스/3/42895. N으로 표현/N으로 표현.swift
import Foundation

func solution(_ N: Int, _ number: Int) -> Int {
    if N == number { return 1 }
    
    var dp = Array(repeating: Set<Int>(), count: 9) // dp[i]는 i번 사용해서 만들 수 있는 숫자 집합
    
    for i in 1...8 {
        let basicNumber = Int(String(repeating: "\(N)", count: i))! // N, NN, NNN...
        dp[i].insert(basicNumber)
        
        for j in 1..<i {
            for num1 in dp[j] {
                for num2 in dp[i - j] {
                    dp[i].insert(num1 + num2) // 덧셈
                    dp[i].insert(num1 - num2) // 뺄셈
                    dp[i].insert(num1 * num2) // 곱셈
                    if num2 != 0 { dp[i].insert(num1 / num2) } // 나눗셈 (0 나누기 방지)
                }
            }
        }
        
        if dp[i].contains(number) {
            return i
        }
    }
    
    return -1
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        if (N == number) return 1;

        Set<Integer>[] dp = new HashSet[9];
        
        for (int i = 1; i <= 8; i++) {
            dp[i] = new HashSet<>();

            int repeated = 0;
            for (int j = 0; j < i; j++) {
                repeated = repeated * 10 + N;
            }
            dp[i].add(repeated);

            for (int j = 1; j < i; j++) {
                for (int a : dp[j]) {
                    for (int b : dp[i - j]) {
                        dp[i].add(a + b);
                        dp[i].add(a - b);
                        dp[i].add(a * b);
                        if (b != 0) dp[i].add(a / b);
                    }
                }
            }

            if (dp[i].contains(number)) return i;
        }

        return -1;
    }
}
